src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1311,7 +1311,7 @@ public Service getService(String type, String algorithm) {
     // re-use will occur e.g. as the framework traverses the provider
     // list and queries each provider with the same values until it finds
     // a matching service
-    private static volatile ServiceKey previousKey =
+    private transient ServiceKey previousKey =
```

Comment:
`transient` is for serialization and I don't think this should affect performance here.  Have you tried just `private ServiceKey ...`?